grace/models/train.py

```diff
@@ -29,33 +48,62 @@ def train_model(
         # weight_decay=5e-4),
     )
 
-    criterion = torch.nn.CrossEntropyLoss()
+    node_criterion = torch.nn.CrossEntropyLoss(
+        ignore_index=Annotation.UNKNOWN, size_average=True
+    )
 
     def train():
         model.train()
 
         for data in train_loader:
-            out = model(data.x, data.edge_index, data.batch)
-            loss = criterion(out, data.y)
+            out_x, out_embedding = model(data.x, data.edge_index, data.batch)
+            loss_node = node_criterion(out_x, data.y)
+            loss_edge = edge_criterion(
+                out_embedding, data.edge_label, data.edge_index, masked_class
+            )
+
+            loss = loss_node + loss_edge
+
             loss.backward()
             optimizer.step()
             optimizer.zero_grad()
 
+        return loss, loss_node, loss_edge
+
     def test(loader):
+        """Evaluates the GCN on node classification."""
         model.eval()
 
-        correct = 0
+        correct_nodes = 0
+        correct_edges = 0
+        num_edges = 0
+
         for data in loader:
-            out = model(data.x, data.edge_index, data.batch)
-            pred = out.argmax(dim=1)  # Use the class with highest probability.
-            correct += int((pred == data.y).sum())
-        return correct / len(loader.dataset)
+            out_x, out_embedding = model(data.x, data.edge_index, data.batch)
+
+            # node
+            pred = out_x.argmax(
+                dim=1
+            )  # Use the class with highest probability.
+            correct_nodes += int((pred == data.y).sum())
+
+            # edge
+            src, dst = data.edge_index
+            edge_score = (out_embedding[src] * out_embedding[dst]).sum(dim=-1)
+            num_edges += edge_score.size(dim=0)
+            correct_edges += int((torch.round(pred) == data.edge_label).sum())
```

Comment:
What is the function of this `round`? Have to admit I'm a little confused by this, because it seems like this is for a binary classification - whereas, we may have multiple edge classes 😕

Reply:
Is this binary because you're ignoring the `UNKNOWN` class?

Reply:
I see, this works because TP and TN are 1 and 0 in the `Annotation` enum and the unknown is ignored?

Reply:
Yup! That's why it works.
Though now that you've mentioned it, it doesn't work for if we add more edge classes than just TP and TN

Reply:
OK - I think that's fine for now. I would just open an issue to make sure that we don't forget this.

Reply:
@quantumjot Do you reckon this requires modification for now?

Reply:
Leave it for now, but open an issue once this is merged